CMakeLists.txt

```diff
@@ -699,6 +699,13 @@ if(CONFIG_APPLICATION_MEMORY)
   list(APPEND LINKER_SCRIPT_DEFINES ${def})
 endif() # CONFIG_APPLICATION_MEMORY
 
+if (CONFIG_CODE_DATA_RELOCATION)
+  set(CODE_RELOCATION_DEP code_relocation_source_lib)
+  # This is needed to force the .common type to .bss type
+  # Without which the common variables cant be placed in the required section
+  zephyr_compile_options(-fno-common)
```

Comment:
Why can't the common section be relocated just like .bss is?

Could you elaborate on what is special about it in this comment?

Reply:
> Common symbols only exist in object files, not in executable files. The idea of COMMON symbols in Unix is to allow multiple external definitions of a same variable (in different compilation units) under a single common symbol under certain conditions.

Ah, I see, so it won't exist in the executable. I wonder how this will affect variables declared with extern.

Reply:
>I think we should apply no-common either everywhere

That is quite outside the scope of this PR. Can do it as a sequel to this PR.

>If there exist two non-extern declarations of the same variable 

This is a feature of the GCC itself and the only reason i am restricting it is because this PR doesn't work without it.

Reply:
> That is quite outside the scope of this PR.

I'm not sure it is.

Imagine that in dozens of places, we are relying on the existence of common symbols.

This would mean that often, when a user enables CONFIG_CODE_RELOCATION, he unexpectedly gets a linker error. Essentially, CONFIG_CODE_RELOCATION is not well-integrated with the rest of Zephyr. This would make this feature unstable, and therefore not
something we want to present to users.

But we don't know if this is the case. Perhaps common symbols are only relied upon in a handful of situations. And these situations can easily be ported to not rely on common symbols. If so, it is best if we port all of these cases by always applying no-common. So that users do not have to discover and report them for us.

In either case. We don't want to apply no-common only when CONFIG_CODE_RELOCATION is enabled.

Essentially, CODE_RELOCATION requires that the C language behaves in a certain way, and for consistency, and maintainability reasons, we want to have it always behave the same.

Reply:
This is a breaking change. What if someone is relying on this feature outside the zephyr tree?. 
If someone is enabling this code relocation feature, he knows what he is getting himself into. But enabling globally by default, I am skeptical.
@nashif Can you please say if it is okay to enable this flag globally(-fno-common)?

Reply:
> What if someone is relying on this feature outside the zephyr tree?

If someone relies on it, and needs to continue to rely on it. Then we should find a way to avoid enabling -fno-common, or else CODE_RELOCATION should not be in-tree.

I am against having the semantics of COMMON be different depending on what features are enabled. The language must always behave the same.

Reply:
Enabled it globally.

Reply:
@SebastianBoe please re-review

Reply:
Done. https://github.com/zephyrproject-rtos/zephyr/pull/10892#discussion_r236717462